README.md
# Personal Identification Number validation

[![Latest Version](https://img.shields.io/packagist/v/vimishor/cnp-php.svg?style=flat-square)](https://packagist.org/packages/vimishor/cnp-php)
[![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](LICENSE)
[![Build Status](https://img.shields.io/travis/vimishor/cnp-php/develop.svg?style=flat-square)](https://travis-ci.org/vimishor/cnp-php)
[![Coverage Status](https://img.shields.io/scrutinizer/coverage/g/vimishor/cnp-php.svg?style=flat-square)](https://scrutinizer-ci.com/g/vimishor/cnp-php/?branch=develop)
[![Code quality](https://img.shields.io/scrutinizer/g/vimishor/cnp-php.svg?style=flat-square)](https://scrutinizer-ci.com/g/vimishor/cnp-php/?branch=develop)
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fvimishor%2Fcnp-php.svg?type=shield)](https://app.fossa.io/projects/git%2Bgithub.com%2Fvimishor%2Fcnp-php?ref=badge_shield)

PHP implementation of [Personal Identification Number specification - draft 0.1](https://github.com/vimishor/cnp-spec), 
in order to validate Personal Identification Number of Romanian citizens and residents.

## Install

Via Composer

``` bash
$ composer require vimishor/cnp-php
```

## Usage

``` php
use Gentle\Embeddable\Date;
use Vimishor\Cnp\Checksum;
use Vimishor\Cnp\Cnp;
use Vimishor\Cnp\County;
use Vimishor\Cnp\Gender;
use Vimishor\Cnp\Serial;

try {
    $cnp = Cnp::fromString('2791219470034');
} catch (\Vimishor\Cnp\Exception\InvalidCnpException $e) {
    // do something
}

// you can also instantiate everything yourself
try {
    $cnp = new Cnp(
        new Gender(2), Date::fromString('1979-12-19T19:10:23+00:00'), new County(47), new Serial(003), new Checksum(4)
    );
} catch (\Vimishor\Cnp\Exception\InvalidCnpException $e) {
    // do something
}
```

## Testing

``` bash
$ make test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

For any security related issues, please email send an email at [alex@gentle.ro][maintainer-pgp] instead of using the issue tracker.

## License

Licensed under the MIT License - see the LICENSE file for details.

[maintainer-pgp]: https://keybase.io/vimishor/key.asc


[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fvimishor%2Fcnp-php.svg?type=large)](https://app.fossa.io/projects/git%2Bgithub.com%2Fvimishor%2Fcnp-php?ref=badge_large)

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.  
This project adheres to [Semantic Versioning](http://semver.org/).

## 0.2.0 / 2019-10-28

  - Removed support for HHVM
  - Added support for PHP 7.2 & 7.3

## 0.1.0 / 2017-08-14

  - First public release


tests/CnpTest.php
<?php
/**
 * Copyright 2017 Alexandru Guzinschi <alex@gentle.ro>
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
namespace Vimishor\Cnp\Test;

use Gentle\Embeddable\Date;
use Vimishor\Cnp\Checksum;
use Vimishor\Cnp\Cnp;
use Vimishor\Cnp\County;
use Vimishor\Cnp\Gender;
use Vimishor\Cnp\Serial;

/**
 * @author Alexandru Guzinschi <alex@gentle.ro>
 */
class CnpTest extends TestCase
{
    /**
     * @param string|int $value
     *
     * @dataProvider invalidCnpProvider
     * @expectedException \Vimishor\Cnp\Exception\InvalidCnpException
     */
    public function testInvalidCnp($value)
    {
        Cnp::fromString($value);
    }

    /**
     * @dataProvider validCnpProvider
     */
    public function testResident($gender, $date, $county, $serial, $controlNumber)
    {
        $cnp = new Cnp($gender, $date, $county, $serial, $controlNumber);

        if (in_array((string)$gender, ['7', '8'], true)) {
            $this->assertTrue($cnp->isResident());
        }
    }

    /**
     * @dataProvider validCnpProvider
     */
    public function testEqualityByValue($gender, $date, $county, $serial, $controlNumber)
    {
        $cnp1 = new Cnp($gender, $date, $county, $serial, $controlNumber);
        $cnp2 = new Cnp($gender, $date, $county, $serial, $controlNumber);

        $this->assertTrue($cnp1->equals($cnp2));
        $this->assertNotSame($cnp1, $cnp2);
    }

    public function validCnpProvider()
    {
        return [
            // gender=f, resident=no, born=19-12-1979, county=Bucuresti/Sectorul 7
            [new Gender(2), Date::fromString('1979-12-19T19:10:23+00:00'), new County(47), new Serial(003), new Checksum(4)],

            // gender=f, resident=no, born=21-02-1864, county=Bucuresti/Sectorul 8
            [new Gender(4), Date::fromString('1864-02-21T19:10:23+00:00'), new County(48), new Serial('503'), new Checksum(1)],

            // gender=m, resident=no, born=21-02-1964, county=Bucuresti/Sectorul 8
            [new Gender(1), Date::fromString('1964-02-21T19:10:23+00:00'), new County(48), new Serial('003'), new Checksum(5)],

            // gender=m, resident=no, born=14-08-1964, county=Braila
            [new Gender(1), Date::fromString('1964-08-14T19:10:23+00:00'), new County(9), new Serial('003'), new Checksum(9)],

            // gender=f, resident=yes, born=21-04-1981, county=Bucuresti/Sectorul 1
            [new Gender(8), Date::fromString('1981-04-21T19:10:23+00:00'), new County(41), new Serial('003'), new Checksum(2)],

            // gender=m, resident=yes, born=23-04-1981, county=Bucuresti/Sectorul 1
            [new Gender(7), Date::fromString('1981-04-23T19:10:23+00:00'), new County(42), new Serial('003'), new Checksum(3)],

            // gender=m, resident=no, born=04-11-2003, county=Arad
            [new Gender(5), Date::fromString('2003-11-04T19:10:23+00:00'), new County('02'), new Serial('415'), new Checksum(9)],

            // gender=m, resident=no, born=24-04-2022, county=Covasna
            [new Gender(5), Date::fromString('2022-04-24T19:10:23+00:00'), new County(14), new Serial('415'), new Checksum(3)],
        ];
    }

    public function invalidCnpProvider()
    {
        return [
            ['484092141038'],   // to short
            ['48409214100381'], // to long
            ['5020315254123'],  // invalid checksum
            ['9020315254129'],  // unknown gender
            ['2920421473211'],  // district 7 after December 19, 1979
            ['7920421483218'],  // resident with district 8
            ['3920317601119'],  // unknown county
            ['6160334271111'],  // invalid day: 34
            ['2940231391116'],  // invalid day: Feb, 31
            ['1921116170008'],  // invalid serial
        ];
    }
}


tests/TestCase.php
<?php
/**
 * Copyright 2017 Alexandru Guzinschi <alex@gentle.ro>
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
namespace Vimishor\Cnp\Test;

use PHPUnit_Framework_TestCase;

/**
 * @author Alexandru Guzinschi <alex@gentle.ro>
 */
abstract class TestCase extends PHPUnit_Framework_TestCase
{
    private static $cloneCache = [];

    public function setUp()
    {
        $this->autoCheckCloning(
            str_replace(array('Test\\', 'Test'), '', get_class($this))
        );
    }

    /**
     * Make sure the `__clone` method is not part of the public API.
     *
     * @param string $class
     */
    protected function assertObjectCantBeCloned($class)
    {
        $ref = new \ReflectionClass($class);
        if (false !== $ref->isCloneable()) {
            $this->fail(sprintf('"%s" should not be cloneable.', $class));
        }
    }

    /**
     * Automatically check each class if is cloneable.
     *
     * This is used as a method to enforce no cloning rule for any
     * future objects.
     *
     * @param string $class
     */
    private function autoCheckCloning($class)
    {
        if (!in_array($class, self::$cloneCache, false)) {
            self::$cloneCache[] = $class;
            $this->assertObjectCantBeCloned($class);
        }
    }
}


tests/CountyTest.php
<?php
/**
 * Copyright 2017 Alexandru Guzinschi <alex@gentle.ro>
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
namespace Vimishor\Cnp\Test;

use Vimishor\Cnp\County;

/**
 * @author Alexandru Guzinschi <alex@gentle.ro>
 */
class CountyTest extends TestCase
{
    /**
     * @param string|int $value
     * @return void
     *
     * @dataProvider validCountyProvider
     */
    public function testInstantiateSuccess($value)
    {
        $county = new County($value);
        $this->assertInstanceOf(County::class, $county);
    }

    /**
     * value should always have a leading zero
     *
     * @param string|int $value
     * @return void
     *
     * @dataProvider validCountyProvider
     */
    public function testValueNormalization($value)
    {
        $county = new County($value);
        $this->assertSame(2, mb_strlen((string)$county));
        $this->assertEquals(str_pad((string)$value, 2, 0, STR_PAD_LEFT), $county->__toString());
    }

    /**
     * @param string|int $value
     *
     * @dataProvider invalidCountyProvider
     * @expectedException \InvalidArgumentException
     */
    public function testInstantiateError($value)
    {
        new County($value);
    }

    /**
     * @param string|int $value
     *
     * @dataProvider validCountyProvider
     */
    public function testEqualityByValue($value)
    {
        $county1 = new County($value);
        $county2 = new County($value);

        $this->assertTrue($county1->equals($county2));
        $this->assertNotSame($county1, $county2);
    }

    /**
     * @param string|int $value
     *
     * @dataProvider validCountyProvider
     */
    public function testEqualityByCountyName($value)
    {
        $county1 = new County($value);
        $county2 = new County($value);

        $this->assertEquals($county1->getName(), $county2->getName());
        $this->assertNotSame($county1, $county2);
    }

    public function validCountyProvider()
    {
        return [
            [1], [02], ['02'], ['47'], [36], ['24']
        ];
    }

    public function invalidCountyProvider()
    {
        return [
            [new \stdClass()], ['0,1'], ['32,43'], ['21.4'], ['string'], ['60'], [60], [[1]], [[02]]
        ];
    }
}


src/Cnp.php
<?php
/**
 * Copyright 2017 Alexandru Guzinschi <alex@gentle.ro>
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
namespace Vimishor\Cnp;

use Gentle\Embeddable\Embeddable;
use Gentle\Embeddable\Date;
use Vimishor\Cnp\Exception\InvalidCnpException;

/**
 * @author Alexandru Guzinschi <alex@gentle.ro>
 */
final class Cnp extends Embeddable
{
    /** @var Gender */
    private $gender;

    /** @var Date */
    private $birthday;

    /** @var County */
    private $county;

    /** @var Serial */
    private $serial;

    /** @var Checksum */
    private $checksum;

    /**
     * Build a {@link CNP} instance from a CNP number.
     *
     * @static
     * @access public
     * @param  string $number A CNP number.
     * @return self
     *
     * @throws InvalidCnpException
     */
    public static function fromString($number)
    {
        if (13 !== mb_strlen($number)) {
            throw new InvalidCnpException(
                'Invalid CNP length.',
                0,
                new \LengthException('Expected a string number of 13 characters.')
            );
        }

        try {
            return new self(
                new Gender(self::getValue($number, 0)),
                new Date(
                    new Date\Year('19'.self::getValue($number, 1).self::getValue($number, 2)),
                    new Date\Month(self::getValue($number, 3).self::getValue($number, 4)),
                    new Date\Day(self::getValue($number, 5).self::getValue($number, 6))
                ),
                new County(self::getValue($number, 7).self::getValue($number, 8)),
                new Serial(self::getValue($number, 9).self::getValue($number, 10).self::getValue($number, 11)),
                new Checksum(self::getValue($number, 12))
            );
        } catch (\Exception $e) {
            throw new InvalidCnpException($e->getMessage(), $e->getCode(), $e);
        }
    }

    /**
     * @param Gender   $gender
     * @param Date     $birthday
     * @param County   $county
     * @param Serial   $serial
     * @param Checksum $checksum
     *
     * @throws InvalidCnpException
     */
    public function __construct(Gender $gender, Date $birthday, County $county, Serial $serial, Checksum $checksum)
    {
        $this->gender = $gender;
        $this->serial = $serial;
        $this->checksum = $checksum;

        try {
            $this->setBirthday($birthday);
            $this->setCounty($county);
        } catch (\DomainException $e) {
            throw new InvalidCnpException('Provided number is not a valid CNP.', $e->getCode(), $e);
        }

        $this->validate((string)$this);
    }

    /**
     * @access public
     * @return Gender
     */
    public function getGender()
    {
        return $this->gender;
    }

    /**
     * @access public
     * @return Date
     */
    public function getBirthday()
    {
        return $this->birthday;
    }

    /**
     * @access public
     * @return County
     */
    public function getCounty()
    {
        return $this->county;
    }

    /**
     * @access public
     * @return Serial
     */
    public function getSerial()
    {
        return $this->serial;
    }

    /**
     * @access public
     * @return Checksum
     */
    public function getChecksum()
    {
        return $this->checksum;
    }

    /**
     * @access public
     * @return bool
     */
    public function isResident()
    {
        return in_array(sprintf('%s', $this->getGender()), ['7', '8'], false);
    }

    /**
     * {@inheritDoc}
     */
    public function equals(Embeddable $object)
    {
        return get_class($object) === 'Vimishor\Cnp\Cnp' && (string)$this === (string)$object;
    }

    /**
     * {@inheritDoc}
     */
    public function __toString()
    {
        return (string)$this->getGender().
            (string)$this->getBirthday()->asDateTime()->format('ymd').
            (string)$this->getCounty().
            (string)$this->getSerial().
            (string)$this->getChecksum();
    }

    /**
     * Rebuild a date object with the year changed.
     *
     * Year prefix is changed according to gender.
     *
     * @access private
     * @param  Date $birthday
     * @return $this
     */
    private function setBirthday(Date $birthday)
    {
        $year   = (string)$birthday->getYear();
        $prefix = $this->getYearPrefix($this->getGender());

        /** @noinspection ExceptionsAnnotatingAndHandlingInspection */
        $date = new Date(
            new Date\Year($prefix . self::getValue($year, 2) . self::getValue($year, 3)),
            new Date\Month((string)$birthday->getMonth()),
            new Date\Day((string)$birthday->getDay())
        );

        $this->birthday = $date;

        return $this;
    }

    /**
     * @access private
     * @param  County $county
     * @return $this
     *
     * @throws \DomainException
     */
    private function setCounty(County $county)
    {
        // Allow usage of Bucharest's district 7 and 8 only for dates prior to December 19, 1979.
        if (in_array((string)$county, ['47', '48'], false) &&
            strtotime($this->getBirthday()) > strtotime(Date::fromString('1979-12-19T23:59:23+00:00'))
        ) {
            $district = (string)$county;
            throw new \DomainException(
                sprintf("Bucharest's district %d doesn't exist anymore since 1979.", $district[1])
            );
        }

        $this->county = $county;

        return $this;
    }

    /**
     * Extract element from string at given position.
     *
     * @access private
     * @param  string $from
     * @param  int    $position
     * @return string
     */
    private static function getValue($from, $position)
    {
        $arr = str_split($from);

        return (string)$arr[$position];
    }

    /**
     * @access private
     * @param  Gender $gender
     * @return int
     */
    private function getYearPrefix(Gender $gender)
    {
        $prefix = 0;

        switch ((int)sprintf('%s', $gender)) {
            case 1:
            case 2:
            case 7:
            case 8:
                $prefix = 19;
                break;
            case 3:
            case 4:
                $prefix = 18;
                break;
            case 5:
            case 6:
                $prefix = 20;
                break;
        } // @codeCoverageIgnore

        return $prefix;
    }

    /**
     * Calculate checksum and compare with the one provided in order to validate
     *
     * @access private
     * @param  string $cnp
     * @return void
     *
     * @throws \Vimishor\Cnp\Exception\InvalidCnpException
     */
    private function validate($cnp)
    {
        // calculate control number and compare in order to validate
        if ((int)sprintf('%s', $this->getChecksum()) !== (int)$this->calculateChecksum($cnp)) {
            throw new InvalidCnpException('Provided number is not a valid CNP.');
        }
    }

    /**
     * @access private
     * @param  string $cnp
     * @return int
     */
    private function calculateChecksum($cnp)
    {
        $sum    = 0;

        for ($i = 0; $i <= 11; ++$i) {
            $sum += (int) self::getValue((string) $cnp, $i) * (int) self::getValue('279146358279', $i);
        }

        $sum %= 11;

        // Rule says that 10 should be interpreted as 1
        return (10 === $sum) ? 1 : $sum;
    }
}


src/Gender.php
<?php
/**
 * Copyright 2017 Alexandru Guzinschi <alex@gentle.ro>
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
namespace Vimishor\Cnp;

use Gentle\Embeddable\Embeddable;

/**
 * @author Alexandru Guzinschi <alex@gentle.ro>
 */
final class Gender extends Embeddable
{
    /**
     * @param string|int $gender Odd for male, even for female.
     *
     * @throws \InvalidArgumentException
     */
    public function __construct($gender)
    {
        $gender = is_string($gender) ? (int)$gender : $gender;

        if (!is_int($gender) || !in_array($gender, range(1, 8), false)) {
            throw new \InvalidArgumentException('Unknown gender');
        }

        $this->value = (string)$gender;
    }

    /**
     * @access public
     * @return bool
     */
    public function isMale()
    {
        return (int)$this->value % 2 !== 0;
    }

    /**
     * {@inheritDoc}
     */
    public function equals(Embeddable $object)
    {
        return get_class($object) === 'Vimishor\Cnp\Gender' && $this->value === (string)$object;
    }
}


src/Serial.php
<?php
/**
 * Copyright 2017 Alexandru Guzinschi <alex@gentle.ro>
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
namespace Vimishor\Cnp;

use Gentle\Embeddable\Embeddable;

/**
 * @author Alexandru Guzinschi <alex@gentle.ro>
 */
final class Serial extends Embeddable
{
    /**
     * @param string|int $number
     *
     * @throws \InvalidArgumentException
     * @throws \OutOfRangeException
     */
    public function __construct($number)
    {
        if (is_int($number)) {
            $number = (string)$number;
        }

        if (!is_string($number) || !ctype_digit($number)) {
            throw new \InvalidArgumentException(sprintf('Expected string or integer and got %s', gettype($number)));
        }

        if ((int)$number > 999 || (int)$number < 1) {
            throw new \OutOfRangeException(sprintf('Expected a number between 1 and 999 (got %d)', (int)$number));
        }

        $this->value = str_pad((string)$number, 3, 0, STR_PAD_LEFT);
    }

    /**
     * {@inheritDoc}
     */
    public function equals(Embeddable $object)
    {
        return get_class($object) === 'Vimishor\Cnp\Serial' && $this->value === (string)$object;
    }
}


src/Checksum.php
<?php
/**
 * Copyright 2017 Alexandru Guzinschi <alex@gentle.ro>
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
namespace Vimishor\Cnp;

use Gentle\Embeddable\Embeddable;

/**
 * @author Alexandru Guzinschi <alex@gentle.ro>
 */
final class Checksum extends Embeddable
{
    /**
     * @param string|int $number
     *
     * @throws \InvalidArgumentException
     */
    public function __construct($number)
    {
        $number = is_int($number) ? (string)$number : $number;

        if (!is_string($number) || !ctype_digit($number) || mb_strlen($number) !== 1) {
            throw new \InvalidArgumentException('Expected a string of length 1');
        }

        $this->value = $number;
    }

    /**
     * {@inheritDoc}
     */
    public function equals(Embeddable $object)
    {
        return get_class($object) === 'Vimishor\Cnp\Checksum' && $this->value === (string)$object;
    }
}


src/County.php
<?php
/**
 * Copyright 2017 Alexandru Guzinschi <alex@gentle.ro>
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
namespace Vimishor\Cnp;

use Gentle\Embeddable\Embeddable;

/**
 * @author Alexandru Guzinschi <alex@gentle.ro>
 */
final class County extends Embeddable
{
    /** @var array */
    private static $counties = [
        '01' => 'Alba',
        '02' => 'Arad',
        '03' => 'Arges',
        '04' => 'Bacau',
        '05' => 'Bihor',
        '06' => 'Bistrita-Nasaud',
        '07' => 'Botosani',
        '08' => 'Brasov',
        '09' => 'Braila',
        '10' => 'Buzau',
        '11' => 'Caras-Severin',
        '12' => 'Cluj',
        '13' => 'Constanta',
        '14' => 'Covasna',
        '15' => 'Dambovita',
        '16' => 'Dolj',
        '17' => 'Galati',
        '18' => 'Gorj',
        '19' => 'Harghita',
        '20' => 'Hunedoara',
        '21' => 'Ialomita',
        '22' => 'Iasi',
        '23' => 'Ilfov',
        '24' => 'Maramures',
        '25' => 'Mehedinti',
        '26' => 'Mures',
        '27' => 'Neamt',
        '28' => 'Olt',
        '29' => 'Prahova',
        '30' => 'Satu Mare',
        '31' => 'Salaj',
        '32' => 'Sibiu',
        '33' => 'Suceava',
        '34' => 'Teleorman',
        '35' => 'Timis',
        '36' => 'Tulcea',
        '37' => 'Vaslui',
        '38' => 'Valcea',
        '39' => 'Vrancea',
        '40' => 'Bucuresti',
        '41' => 'Bucuresti/Sectorul 1',
        '42' => 'Bucuresti/Sectorul 2',
        '43' => 'Bucuresti/Sectorul 3', // was merged with district 2 in 1979 => district 2
        '44' => 'Bucuresti/Sectorul 4', // district 3 since 1979
        '45' => 'Bucuresti/Sectorul 5', // district 4 since 1979
        '46' => 'Bucuresti/Sectorul 6', // district 5 since 1979
        '47' => 'Bucuresti/Sectorul 7', // district 6 since 1979
        '48' => 'Bucuresti/Sectorul 8', // was merged with district 1 in 1979 => district 1
        '51' => 'Calarasi',
        '52' => 'Giurgiu'
    ];

    /**
     * @param string|int $county
     *
     * @throws \InvalidArgumentException
     */
    public function __construct($county)
    {
        if (is_int($county)) {
            $county = (string)$county;
        }

        if (!is_string($county) && !ctype_digit($county)) {
            throw new \InvalidArgumentException(sprintf('Expected string or integer and got %s', gettype($county)));
        }

        $county = str_pad((string)$county, 2, 0, STR_PAD_LEFT);

        if (!array_key_exists($county, self::$counties)) {
            throw new \InvalidArgumentException(sprintf('%s is not a known county.', $county));
        }

        $this->value = $county;
    }

    /**
     * @access public
     * @return string
     */
    public function getName()
    {
        return self::$counties[$this->value];
    }

    /**
     * {@inheritDoc}
     */
    public function equals(Embeddable $object)
    {
        return get_class($object) === 'Vimishor\Cnp\County' && $this->value === (string)$object;
    }
}
